### DISCUSSION TITLE
Best way to integrate a websocket?

### DISCUSSION THREAD

**Discussion by tonymaynard97** [pmndrs/zustand#1651] [Q&A]
Just started reading and playing around with zustand, I currently have some slices in my store but I'd like to introduce a new slice for some live data. Whats the best approach to do this with zustand? Shall I store reference to an initialised web socket and then dispatch some actions to send messages to the server? Or shall I create a component which handles the WebSocket connection and then reads messages from an array in the store and pops them off.

**Comment by 0xbuidler** [comment]
Seems like a lot of people would need some insights, i'm just bumping this.

**Comment by itekhi** [comment]
I would like to know that too... Didn't found any info on the internet.
I'm currently working on react native app that needs a persistent WS connection (to reconnect on app reopen).
I will try to store initialized WS connection in the store, and an ID of connection in the `persist` (with `partialize` to not store the socket), then in `onRehydrateStorage` if there's an `ID` and no socket - reconnect.

Any ideas on what problems could appear?

**Comment by kulame** [comment]
I also want to know

**Comment by tonymaynard97** [comment]
So for others benefit I resulted in initialising the WebSocket connection within the store, then in the components dispatch a message which hits a reducer within the store that has reference to the web socket and then sends the message natively over web socket. Unfortunately I cannot provide my code but its similar to something below:

```javascript
export function store(set, get) {
  let ws: WebSocket

  const openHandler = () => {...}
    const messageHandler = () => {
    // Call settters here 
    }
      const closeHandler = () => {
    console.log('Closed WS')
    set({ subscribed: {} })
    ws = new WebSocket(URL)

    ws.addEventListener('open', openHandler)
    ws.addEventListener('message', messageHandler)
    ws.addEventListener('close', closeHandler)
  }
  closeHandler()
  
    return {
    actions: {
      dispatch: (args) =>
        set((state) => reducer(state, args, ws, get)),
    },
  }
}

```
